concent_api/core/payments/bankster.py

```diff
@@ -71,96 +81,165 @@ def claim_deposit(
+    if is_claim_against_provider:
+        provider_deposit = service.get_deposit_value(client_eth_address=provider_ethereum_address)  # pylint: disable=no-value-for-parameter
+
+    # Bankster puts database locks on DepositAccount objects
+    # that will be used as payers in newly created DatabaseClaims.
+    with transaction.atomic(using='control'):
+        # Bankster sums the amounts of all existing DepositClaims that have the same payer as the one being processed.
+        sum_of_existing_requestor_claims = DepositClaim.objects.filter(
+            payer_deposit_account=requestor_deposit_account
+        ).aggregate(
+            sum_of_existing_claims=Coalesce(Sum('amount'), 0)
+        )
+
+        # If the existing claims against requestor's deposit are greater or equal to his current deposit,
+        # we can't add a new claim.
+        if requestor_deposit <= sum_of_existing_requestor_claims['sum_of_existing_claims']:
+            return (None, None)
+
+        if is_claim_against_provider:
+            # Bankster sums the amounts of all existing DepositClaims where the provider is the payer.
+            sum_of_existing_provider_claims = DepositClaim.objects.filter(
+                payer_deposit_account=provider_deposit_account
+            ).aggregate(
+                sum_of_existing_claims=Coalesce(Sum('amount'), 0)
+            )
+
+            # If the total of existing claims and the current claim is greater or equal to the current deposit,
+            # we can't add a new claim.
+            if provider_deposit <= sum_of_existing_provider_claims['sum_of_existing_claims'] + settings.ADDITIONAL_VERIFICATION_COST:
+                return (requestor_deposit, None)
+
+        # Deposit lock for requestor.
+        claim_against_requestor = DepositClaim(
+            subtask_id=subtask_id,
+            payee_ethereum_address=provider_ethereum_address,
+            amount=subtask_cost,
+            concent_use_case=concent_use_case,
+            payer_deposit_account=requestor_deposit_account,
+        )
+        claim_against_requestor.full_clean()
+        claim_against_requestor.save()
+
+        # Deposit lock for provider.
+        if is_claim_against_provider:
+            claim_against_provider = DepositClaim(
+                subtask_id=subtask_id,
+                payee_ethereum_address=ethereum_public_key_to_address(
+                    settings.CONCENT_ETHEREUM_PUBLIC_KEY
+                ),
+                amount=subtask_cost,
+                concent_use_case=concent_use_case,
+                payer_deposit_account=provider_deposit_account,
+            )
+            claim_against_provider.full_clean()
+            claim_against_provider.save()
+        else:
+            claim_against_provider = None  # type: ignore
+
+    return (claim_against_requestor, claim_against_provider)
+
+
+def finalize_payment(deposit_claim: DepositClaim) -> Optional[str]:
```

Comment:
>> BTW. Why is return type str? This function is called in only one place and return value is not assigned to any variable in production code. But in tests it is always assigned and checked. I think it should not be like this. Return type should be None and tests should be fixed to don't use return value..

> I agree with second part, I don't know where this transaction hash is going to be used later (maybe there is a master plan for this).

The fact that it's not used in the current implementation does not mean that it won't be used in the future. Bankster is an API (remember that it was originally going to be a separate service) and it should be a little more generic than the code that uses it. I actually think that our views are making too many assumptions about that it does. It should be treated as a black box.

About the return value: even if the operation does not fail, there are two possible outcomes. Either there is enough deposit to satisfy the claim (fully or in part) or there isn't. If there isn't, the claim is simply removed. I think that it should not be done silently. The caller should be informed about that and be able to take some action. Right now the view does not really care what happens with the payment after we initiate it but this is not good and should change in the future.

So the function should return something. It could return simply `True` or `False`. Returning the transaction ID serves the same purpose and gives more information so why not? We might want to log it or even return it to the caller in the future. It does not cost us anything to be more flexible and informative here.

> > According to #896 return type should be `str` and it shouldn't be Optional. If it is OK in code please comment in issue (or change).

> About first part, updated in issue +1

I'd like to clarify one thing. The types I'm using in the spec are not as specific as `str`, `bytes`. I simply call it a string and that's on purpose. I do this with other types too (e.g. "decimal" instead of `Decimal` or `float`, "list of" instead of `list` or `tuple`). The actual type is a low level detail. Choose one that is appropriate.

The reason is that there are several different type systems we're using (Python, JSON, SQL, Django models). The same data often has to be moved across several of them and it's hard to describe the type in a way that's both accurate and compatible with all of them. Even when using one, same data can sometimes be stored as `string` or `bytes` depending on various factors. When you try to keep track of it, you often run into small inconsistencies in the spec and trying to get them all right requires a lot of effort. I often find that this all effort is unnecessary and it's simply better to leave it all up to implementation. On top of that, trying to be too precise makes the spec less readable.